File: src/pricing.ts
/** Prices for models that appear in Grok Build.
 *  Grok: official xAI API list prices
 *  Composer: Cursor list prices
 */

export type ModelPrice = {
  id: string
  label: string
  inputPerM: number
  outputPerM: number
  /** Informational only; logs do not expose cache hits reliably */
  cachedInputPerM?: number
  context?: string
  note?: string
}

/** Models that appear in Grok Build (region-dependent). */
export const OFFICIAL_PRICES: ModelPrice[] = [
  {
    id: "grok-4.5",
    label: "Grok 4.5",
    inputPerM: 2.0,
    outputPerM: 6.0,
    cachedInputPerM: 0.5,
    context: "500k",
    note: "xAI API · default in USA (elsewhere often via VPN)",
  },
  {
    id: "grok-build",
    label: "grok-build",
    inputPerM: 1.0,
    outputPerM: 2.0,
    context: "256k",
    note: "xAI Code API · extra option in some regions (e.g. EU)",
  },
  {
    id: "grok-build-0.1",
    label: "grok-build-0.1",
    inputPerM: 1.0,
    outputPerM: 2.0,
    context: "256k",
    note: "Official Code API id (same rates as grok-build)",
  },
  {
    id: "grok-composer-2.5-fast",
    label: "Composer 2.5 Fast",
    inputPerM: 3.0,
    outputPerM: 15.0,
    context: "200k",
    note: "Cursor list price · default outside USA",
  },
  {
    id: "grok-composer-2.5",
    label: "Composer 2.5 Standard",
    inputPerM: 0.5,
    outputPerM: 2.5,
    context: "200k",
    note: "Cursor list price · Standard tier",
  },
]

const byId = new Map(OFFICIAL_PRICES.map((p) => [p.id.toLowerCase(), p]))

const UNPRICED: ModelPrice = {
  id: "unknown",
  label: "Unknown",
  inputPerM: 0,
  outputPerM: 0,
  note: "Not a known Grok Build model / no public list price",
}

export function resolvePrice(modelId: string | null | undefined): ModelPrice {
  if (!modelId) return { ...UNPRICED }

  const key = modelId.toLowerCase().trim()
  const exact = byId.get(key)
  if (exact) return exact

  if (key.startsWith("grok-build")) return byId.get("grok-build")!
  if (key.startsWith("grok-4.5")) return byId.get("grok-4.5")!

  if (key.includes("composer")) {
    if (key.includes("standard") && !key.includes("fast")) {
      return byId.get("grok-composer-2.5")!
    }
    return byId.get("grok-composer-2.5-fast")!
  }

  return { ...UNPRICED, id: modelId, label: modelId }
}

export type CostBreakdown = {
  inputCost: number
  outputCost: number
  totalCost: number
  priced: boolean
  price: ModelPrice
}

export function calcCost(
  modelId: string | null | undefined,
  inputTokens: number,
  outputTokens: number,
): CostBreakdown {
  const price = resolvePrice(modelId)
  const priced = price.inputPerM > 0 || price.outputPerM > 0
  const inputCost = (inputTokens / 1_000_000) * price.inputPerM
  const outputCost = (outputTokens / 1_000_000) * price.outputPerM
  return {
    inputCost,
    outputCost,
    totalCost: inputCost + outputCost,
    priced,
    price,
  }
}

export const PRICING_SOURCE = "https://docs.x.ai/developers/pricing"
export const PRICING_COMPOSER = "https://cursor.com/docs"
export const PRICING_UPDATED = "2026-07"

File: src/pricing.rs
#[derive(Debug, Clone)]
pub struct ModelPrice {
    pub id: &'static str,
    pub label: &'static str,
    pub input_per_m: f64,
    pub output_per_m: f64,
    pub cached_input_per_m: Option<f64>,
    pub context: Option<&'static str>,
    pub note: Option<&'static str>,
}

pub const OFFICIAL_PRICES: &[ModelPrice] = &[
    ModelPrice {
        id: "grok-4.5",
        label: "Grok 4.5",
        input_per_m: 2.0,
        output_per_m: 6.0,
        cached_input_per_m: Some(0.5),
        context: Some("500k"),
        note: Some("xAI API · default in USA (elsewhere often via VPN)"),
    },
    ModelPrice {
        id: "grok-build",
        label: "grok-build",
        input_per_m: 1.0,
        output_per_m: 2.0,
        cached_input_per_m: None,
        context: Some("256k"),
        note: Some("xAI Code API · extra option in some regions (e.g. EU)"),
    },
    ModelPrice {
        id: "grok-build-0.1",
        label: "grok-build-0.1",
        input_per_m: 1.0,
        output_per_m: 2.0,
        cached_input_per_m: None,
        context: Some("256k"),
        note: Some("Official Code API id (same rates as grok-build)"),
    },
    ModelPrice {
        id: "grok-composer-2.5-fast",
        label: "Composer 2.5 Fast",
        input_per_m: 3.0,
        output_per_m: 15.0,
        cached_input_per_m: None,
        context: Some("200k"),
        note: Some("Cursor list price · default outside USA"),
    },
    ModelPrice {
        id: "grok-composer-2.5",
        label: "Composer 2.5 Standard",
        input_per_m: 0.5,
        output_per_m: 2.5,
        cached_input_per_m: None,
        context: Some("200k"),
        note: Some("Cursor list price · Standard tier"),
    },
];

pub const PRICING_SOURCE: &str = "https://docs.x.ai/developers/pricing";
pub const PRICING_UPDATED: &str = "2026-07";

#[derive(Debug, Clone)]
pub struct CostBreakdown {
    pub input_cost: f64,
    pub output_cost: f64,
    pub total_cost: f64,
    pub priced: bool,
    pub price: ModelPriceOwned,
}

#[derive(Debug, Clone)]
pub struct ModelPriceOwned {
    pub id: String,
    #[allow(dead_code)]
    pub label: String,
    pub input_per_m: f64,
    pub output_per_m: f64,
    #[allow(dead_code)]
    pub cached_input_per_m: Option<f64>,
    #[allow(dead_code)]
    pub context: Option<String>,
    #[allow(dead_code)]
    pub note: Option<String>,
}

impl From<&ModelPrice> for ModelPriceOwned {
    fn from(p: &ModelPrice) -> Self {
        Self {
            id: p.id.to_string(),
            label: p.label.to_string(),
            input_per_m: p.input_per_m,
            output_per_m: p.output_per_m,
            cached_input_per_m: p.cached_input_per_m,
            context: p.context.map(|s| s.to_string()),
            note: p.note.map(|s| s.to_string()),
        }
    }
}

fn unpriced(model_id: &str) -> ModelPriceOwned {
    ModelPriceOwned {
        id: model_id.to_string(),
        label: if model_id == "unknown" {
            "Unknown".into()
        } else {
            model_id.to_string()
        },
        input_per_m: 0.0,
        output_per_m: 0.0,
        cached_input_per_m: None,
        context: None,
        note: Some("Not a known Grok Build model / no public list price".into()),
    }
}

pub fn resolve_price(model_id: Option<&str>) -> ModelPriceOwned {
    let Some(model_id) = model_id else {
        return unpriced("unknown");
    };

    let key = model_id.to_lowercase();
    let key = key.trim();

    if let Some(p) = OFFICIAL_PRICES.iter().find(|p| p.id.eq_ignore_ascii_case(key)) {
        return p.into();
    }

    if key.starts_with("grok-build") {
        return OFFICIAL_PRICES
            .iter()
            .find(|p| p.id == "grok-build")
            .unwrap()
            .into();
    }
    if key.starts_with("grok-4.5") {
        return OFFICIAL_PRICES
            .iter()
            .find(|p| p.id == "grok-4.5")
            .unwrap()
            .into();
    }
    if key.contains("composer") {
        if key.contains("standard") && !key.contains("fast") {
            return OFFICIAL_PRICES
                .iter()
                .find(|p| p.id == "grok-composer-2.5")
                .unwrap()
                .into();
        }
        return OFFICIAL_PRICES
            .iter()
            .find(|p| p.id == "grok-composer-2.5-fast")
            .unwrap()
            .into();
    }

    unpriced(model_id)
}

pub fn calc_cost(model_id: Option<&str>, input_tokens: u64, output_tokens: u64) -> CostBreakdown {
    let price = resolve_price(model_id);
    let priced = price.input_per_m > 0.0 || price.output_per_m > 0.0;
    let input_cost = (input_tokens as f64 / 1_000_000.0) * price.input_per_m;
    let output_cost = (output_tokens as f64 / 1_000_000.0) * price.output_per_m;
    CostBreakdown {
        input_cost,
        output_cost,
        total_cost: input_cost + output_cost,
        priced,
        price,
    }
}
